west/cmd/git.py

```diff
@@ -0,0 +1,255 @@
+        super().__init__(
+            'status',
+            "Run 'git status' for each project. Extra arguments are passed "
+            "as-is to 'git status'.",
+            accepts_unknown_args=True)
+
+    def do_add_parser(self, parser_adder):
+        return _add_common_git_flags(parser_adder, self)
+
+    def do_run(self, args, user_args):
+        for project in _all_projects(args):
+            if _check_repo(project):
+                _git(project, 'fetch')
+
+                print("=== 'git status' for {} (in {}) ==="
+                      .format(project.name, project.path))
+                _git(project, 'status', extra_args=user_args)
+
+
+def _add_common_git_flags(parser_adder, command):
+    # Adds common command-line flags for the Git-related commands. The manifest
+    # file contains repository information.
+
+    parser = parser_adder.add_parser(
+        command.name,
+        description=command.description)
+
+    parser.add_argument(
+        '-m', '--manifest',
+        dest='manifest',
+        help='path to manifest file (default: <zephyr-base>/manifest/default.yml)')
+
+    # TODO: Make schema optional?
+
+    parser.add_argument(
+        '-s', '--schema',
+        dest='schema',
+        help='path to pykwalify schema for manifest (default: <zephyr-base>/manifest/default.yml)')
+
+    return parser
+
+
+# Holds information about a project, taken from the manifest file
+Project = collections.namedtuple('Project', 'name url revision path')
+
+
+def _all_projects(args):
+    # Parses the manifest file, returning a list with Project instances for all
+    # projects. Also verifies the manifest against a pykwalify schema.
+
+    if (not args.manifest or not args.schema) and not args.zephyr_base:
+        log.die('Zephyr base directory not specified (via --zephyr-base or ZEPHYR_BASE)')
+
+    if args.schema:
+        schema_filename = args.schema
+    else:
+        schema_filename = os.path.join(
+            args.zephyr_base, 'manifest', 'schema.yml')
+
+    if args.manifest:
+        manifest_filename = args.manifest
+    else:
+        manifest_filename = os.path.join(
+            args.zephyr_base, 'manifest', 'manifest.yml')
+
+
+    # Validate the manifest with pykwalify
+
+    try:
+        pykwalify.core.Core(
+            source_file=manifest_filename, schema_files=[schema_filename]
+        ).validate()
+    except pykwalify.errors.SchemaError as e:
+        log.die('{} malformed (schema: {}):\n{}'
+                .format(manifest_filename, schema_filename, e))
+
+
+    # Load project information from manifest
+
+    with open(manifest_filename) as f:
+        manifest = yaml.load(f)['manifest']
+
+    projects = []
+
+    # mp = manifest project (dictionary with values parsed from the manifest)
+    for mp in manifest['projects']:
+        # Fill in any missing fields in 'mp' with values from the 'defaults'
+        # dictionary
+        for key, val in manifest['defaults'].items():
```

Comment:
This seems safe for now, but note that the manifest-wide defaults might contain keys which do not apply to projects.

Reply:
Thought it was defined as defaults for projects.

Reply:
It's defaults for the manifest

Reply:
I thought the manifest was a project description. :)

Reply:
Repo lets you manage default values that are related to the behavior of the tool itself there. For example: `<default sync-j="4"/>` uses 4 concurrent TCP connections while syncing. I imagine we may end up with use cases that are similar eventually.

Reply:
Ah, yeah, might have to change the approach at that point, if they get added to the `defaults` dict.